File: python/clock/clock.py
# -*- coding: utf-8 -*-

class Clock:

  def __init__(self, hours, minutes):
    self.minutes = minutes % 60
    self.hours = (hours + (minutes // 60)) % 24

  def __str__(self):
    return u"{:02d}:{:02d}".format(self.hours, self.minutes)

  def __eq__(self, other):
    return (isinstance(other, self.__class__)
      and self.hours == other.hours
      and self.minutes == other.minutes)

  def __ne__(self, other):
    return not self.__eq__(other)

  def add(self, minutes):
    return Clock(self.hours, self.minutes + minutes)

File: lua/clock/clock.lua
local clock = {}
clock.__index = clock

local function normalize(hours, minutes)
  local state = {
    minutes = minutes % 60,
    hours = (hours + math.floor(minutes / 60)) % 24
  }
  return setmetatable(state, clock)
end

function clock.at(hours, minutes)
  return normalize(hours, minutes or 0)
end

function clock.__tostring(state)
  return string.format("%02d:%02d", state.hours, state.minutes)
end

function clock:equals(other)
  return tostring(self) == tostring(other)
end

function clock:minus(minutes)
  return normalize(self.hours, self.minutes - minutes)
end

function clock:plus(minutes)
  return normalize(self.hours, self.minutes + minutes)
end

return clock